services/cannon/src/Cannon/App.hs

```diff
@@ -115,12 +115,22 @@ readLoop ws s = loop
                 send (connection ws) (pong p)
                 loop
             ControlMessage (Close _ _) -> return ()
-            _                          -> reset counter s 0 >> loop
+            _ -> do
+                reset counter s 0
+                sendAppLevelPong
```

Comment:
I feel slightly uneasy about sending a pong for every data message – perhaps we should only pong to messages that say "ping"? Though I can't see any _actual_ problems with sending a pong for every data message.

I also feel uneasy about sending and receiving payloads that are not Protobuf/JSON-encoded (or generally encoded in the same way as other payloads), but dragging Protobuf into this code looks like a mistake.

Reply:
> I feel slightly uneasy about sending a pong for every data message – perhaps we should only pong to messages that say "ping"? Though I can't see any _actual_ problems with sending a pong for every data message.

the reason i chose to ignore the ping message is that it is currently a rather ad hoc string, and i think it's brittle and dangerous to depend on that.  note that we don't expect any other messages but `ping`; everything else from the client is sent via http.  (right?)

> I also feel uneasy about sending and receiving payloads that are not Protobuf/JSON-encoded (or generally encoded in the same way as other payloads), but dragging Protobuf into this code looks like a mistake.

i'm indifferent.  let's look into protobuffing it once it's clear what we want to do.  I still hope there is a way to avoid this PR entirely.

Reply:
> I still hope there is a way to avoid this PR entirely.

Agreed.